File: app/src/main/java/com/twitter/tweetducker/model/JSON.java
package com.twitter.tweetducker.model;

import android.util.Log;

import com.crashlytics.android.Crashlytics;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

public class JSON {

    private static final String TAG = JSON.class.getSimpleName();

    private static User parseUser(JsonNode user) {
        User result = null;

        if (user != null) {
            result = new User();
            result.id = user.get("id").asLong();
            result.name = user.get("name").asText();
            result.screenName = user.get("screen_name").asText();
            result.description = user.get("description").asText();
            result.profileImageUrlHttps = user.get("profile_image_url_https").asText();
        }

        return result;
    }

    private static Timeline parseTimeline(JsonNode timeline) {
        Timeline result = null;

        if (timeline != null) {
            result = new Timeline();
            result.name = timeline.get("name").asText();
            result.description = timeline.get("description").asText();
            result.collectionUrl = timeline.get("collection_url").asText();
            result.visibility = timeline.get("visibility").asText();
        }

        return result;
    }

    public static CollectionsList parseCollectionsList(String json, long userId) {
        CollectionsList result = null;

        try {
            ObjectMapper mapper = new ObjectMapper();
            JsonNode root = mapper.readValue(json, JsonNode.class);

            if (root != null) {
                result = new CollectionsList();

                JsonNode user = root.path("objects").path("users").path("" + userId);
                if (!user.isMissingNode()) {
                    result.user = parseUser(user);
                }

                JsonNode timelines = root.path("objects").path("timelines");
                if (!timelines.isMissingNode()) {
                    List<Timeline> tmp = new LinkedList<>();
                    for (JsonNode timeline : timelines) {
                        tmp.add(parseTimeline(timeline));
                    }

                    result.timelines = tmp;
                }
            }
        } catch (IOException ioe) {
            Crashlytics.logException(ioe);
            Log.d(TAG, "IOException parsing JSON.", ioe);
        }

        return result;
    }
}

File: app/src/main/kotlin/com/twitter/tweetducker/JSON.kt
package com.twitter.tweetducker

import com.crashlytics.android.Crashlytics
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper

import java.io.IOException

import kotlin.comparisons.compareBy

object JSON {

    private fun parseUser(user: JsonNode): User = User(
            user.get("id").asLong(),
            user.get("name").asText(),
            user.get("screen_name").asText(),
            user.get("description").asText(),
            user.get("profile_image_url_https").asText()
    )

    private fun parseTimeline(timeline: JsonNode): Timeline = Timeline(
            timeline.get("name").asText(),
            timeline.get("description").asText(),
            timeline.get("collection_url").asText(),
            timeline.get("visibility").asText()
    )

    fun parseCollectionsList(json: String, userId: Long): CollectionsList? {
        try {
            val mapper = ObjectMapper()
            val root = mapper.readValue(json, JsonNode::class.java)

            root?.let {
                val userJson = root.path("objects").path("users").path("${userId}")
                val timelinesJson = root.path("objects").path("timelines")

                if (!userJson.isMissingNode && !timelinesJson.isMissingNode) {
                    val user = parseUser(userJson)

                    val timelines = mutableListOf<Timeline>()
                    for (timeline in timelinesJson) {
                        timelines.add(parseTimeline(timeline))
                    }

                    val sortedTimelines = timelines.sortedWith(compareBy { it.name.toLowerCase() })

                    return CollectionsList(user, sortedTimelines)
                }
            }
        } catch (ioe: IOException) {
            Crashlytics.logException(ioe)
        }

        return null
    }
}
